tasks/fevm/trace/task.go

```diff
@@ -0,0 +1,199 @@
+		if err != nil {
+			return fmt.Errorf("getting messages executions for tipset: %w", err)
+		}
+		return nil
+	})
+
+	var getActorCode func(ctx context.Context, a address.Address) (cid.Cid, bool)
+	grp.Go(func() error {
+		var err error
+		getActorCode, err = util.MakeGetActorCodeFunc(ctx, t.node.Store(), current, executed)
+		if err != nil {
+			return fmt.Errorf("failed to make actor code query function: %w", err)
+		}
+		return nil
+	})
+
+	report := &visormodel.ProcessingReport{
+		Height:    int64(current.Height()),
+		StateRoot: current.ParentState().String(),
+	}
+
+	// if either fail, report error and bail
+	if err := grp.Wait(); err != nil {
+		report.ErrorsDetected = err
+		return nil, report, nil
+	}
+
+	var (
+		vmMessageResults = make(fevm.FEVMTraceList, 0)
+	)
+
+	errs := []error{}
+
+	for _, parentMsg := range mex {
+		// Only handle EVM related message
+		if !util.IsEVMAddress(ctx, t.node, parentMsg.Message.From, current.Key()) && !util.IsEVMAddress(ctx, t.node, parentMsg.Message.To, current.Key()) && !(parentMsg.Message.To != builtintypes.EthereumAddressManagerActorAddr) {
+			continue
+		}
+		messageHash, err := ethtypes.EthHashFromCid(parentMsg.Cid)
+		if err != nil {
+			log.Errorf("Error at finding hash: [cid: %v] err: %v", parentMsg.Cid, err)
+			errs = append(errs, err)
+			continue
+		}
+		transaction, err := t.node.EthGetTransactionByHash(ctx, &messageHash)
+		if err != nil {
+			log.Errorf("Error at getting transaction: [hash: %v] err: %v", messageHash, err)
+			errs = append(errs, err)
+			continue
+		}
+
+		if transaction == nil {
+			continue
+		}
+
+		log.Infof("message: %v, %v", parentMsg.Cid, base64.StdEncoding.EncodeToString(parentMsg.Message.Params))
+
+		for _, child := range util.GetChildMessagesOf(parentMsg) {
+			toCode, _ := getActorCode(ctx, child.Message.To)
+
+			toActorCode := "<Unknown>"
+			if !toCode.Equals(cid.Undef) {
+				toActorCode = toCode.String()
+			}
+			fromEthAddress := getEthAddress(child.Message.From)
+			toEthAddress := getEthAddress(child.Message.To)
+
+			vmMsg := &fevm.FEVMTrace{
+				Height:           int64(parentMsg.Height),
+				TransactionHash:  transaction.Hash.String(),
+				MessageStateRoot: parentMsg.StateRoot.String(),
+				MessageCid:       parentMsg.Cid.String(),
+				TraceCid:         getMessageTraceCid(child.Message).String(),
+				To:               child.Message.To.String(),
+				From:             child.Message.From.String(),
+				FromEthAddress:   fromEthAddress,
+				ToEthAddress:     toEthAddress,
+				Value:            child.Message.Value.String(),
+				GasUsed:          0,
+				ExitCode:         int64(child.Receipt.ExitCode),
+				ActorCode:        toActorCode,
+				Method:           uint64(child.Message.Method),
+				Index:            child.Index,
+				Params:           base64.StdEncoding.EncodeToString(child.Message.Params),
+				Returns:          base64.StdEncoding.EncodeToString(child.Receipt.Return),
```

Comment:
maybe use `EthBytes` to be consistent with eth?

Reply:
changed